src/modelframe.jl

```diff
@@ -1,261 +1,102 @@
+        res[i] &= !ismissing(el)
     end
+end
 
-    for i_term in 1:n_terms
-        for i_eterm in 1:n_eterms
-            ## only need to check eterms that are included and can be promoted
-            ## (e.g., categorical variables that expand to multiple mm columns)
-            if Bool(trms.factors[i_eterm, i_term]) && is_categorical(df[trms.eterms[i_eterm]])
-                dropped = trms.factors[:,i_term]
-                dropped[i_eterm] = 0
-
-                if dropped ∉ encountered_columns
-                    trms.is_non_redundant[i_eterm, i_term] = true
-                    push!(encountered_columns, dropped)
-                end
-
-            end
-        end
-        ## once we've checked all the eterms in this term, add it to the list
-        ## of encountered terms/columns
-        push!(encountered_columns, view(trms.factors, :, i_term))
+function _nonmissing!(res, col::CategoricalArray{>: Missing})
+    for (i, el) in enumerate(col.refs)
+        res[i] &= el > 0
     end
-
-    return trms.is_non_redundant
 end
 
-const DEFAULT_CONTRASTS = DummyCoding
-
-# get unique values, honoring levels order
-_unique(x::AbstractCategoricalArray) = intersect(levels(x), unique(x))
-_unique(x::AbstractCategoricalArray{T}) where {T>:Missing} =
-    convert(Array{Missings.T(T)}, filter!(!ismissing, intersect(levels(x), unique(x))))
+_select(d::ColumnTable, cols::NTuple{N,Symbol} where N) = NamedTuple{cols}(d)
+_select(d::ColumnTable, cols::Vector{Symbol}) = _select(d, tuple(cols...))
+_filter(d::T, rows) where T<:ColumnTable = T(([col[rows] for col in d]..., ))
 
-function _unique(x::AbstractArray{T}) where T
-    levs = T >: Missing ?
-           convert(Array{Missings.T(T)}, filter!(!ismissing, unique(x))) :
-           unique(x)
-    try; sort!(levs); catch; end
-    return levs
-end
-
-## Set up contrasts:
-## Combine actual DF columns and contrast types if necessary to compute the
-## actual contrasts matrices, levels, and term names (using DummyCoding
-## as the default)
-function evalcontrasts(df::AbstractDataFrame, contrasts::Dict = Dict())
-    evaledContrasts = Dict()
-    for (term, col) in eachcol(df)
-        is_categorical(col) || continue
-        evaledContrasts[term] = ContrastsMatrix(haskey(contrasts, term) ?
-                                                contrasts[term] :
-                                                DEFAULT_CONTRASTS(),
-                                                _unique(col))
-    end
-    return evaledContrasts
-end
+_size(d::ColumnTable) = (length(first(d)), length(d))
+_size(d::ColumnTable, dim::Int) = _size(d)[dim]
 
 ## Default NULL handler.  Others can be added as keyword arguments
-function missing_omit(df::DataFrame)
-    cc = completecases(df)
-    df[cc,:], cc
+function missing_omit(d::T) where T<:ColumnTable
+    nonmissings = trues(_size(d, 1))
+    for col in d
+        _nonmissing!(nonmissings, col)
+    end
+    map(disallowmissing, _filter(d, nonmissings)), nonmissings
```

Comment:
As noted in a previous comment: you could avoid an intermediate allocation if `_filter` allocated a vector with the right element type directly (by calling `Missings.T`) and called `copyto!`.

(Also `_filter` could use a generator to avoid a smaller allocation.)

Reply:
I think I understand in theory how this would work but I can't quite figure out how actually do it.

Reply:
Maybe this (to replace `_filter(...)`)?
```julia
rows = findall(nonmissings)
T(tuple((copyto!(similar(col, Missings.T(eltype(col)), length(rows)),
                 view(col, rows) for col in d)...))
```

(Calling `rows = findall(...)` manually prevents `view` from doing that repeatedly for each column.)

It's too bad we don't have a way to extract the type that `similar` would return, so that we could just call `convert` on that.

Reply:
The problem here is that `T` still has the wrong eltype for the vectors that contained missings.

Reply:
I've incorporated this suggestion with the tweak of using the names-only `NamedTuple{names}()` constructor instead of `T` directly